File: backend/models/training.py
"""
SQLAlchemy models for automated fine-tuning pipeline.
"""

from __future__ import annotations

from sqlalchemy import (
    Column,
    DateTime,
    ForeignKey,
    Integer,
    JSON,
    String,
    Text,
)
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from backend.models import Base


class TrainingDataset(Base):
    """Represents a generated dataset of instruction-response pairs."""

    __tablename__ = "training_datasets"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String, nullable=False)
    description = Column(Text, nullable=True)
    status = Column(String, default="pending", nullable=False, index=True)
    source_documents = Column(JSON, nullable=False)  # list of document ids with metadata
    config = Column(JSON, nullable=True)
    storage_path = Column(String, nullable=True)  # local filesystem path
    artifact_path = Column(String, nullable=True)  # path in MinIO or external storage
    total_examples = Column(Integer, default=0, nullable=False)
    sample_preview = Column(JSON, nullable=True)
    error = Column(Text, nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

    jobs = relationship("TrainingJob", back_populates="dataset")


class TrainingJob(Base):
    """Tracks fine-tuning jobs executed via Unsloth."""

    __tablename__ = "training_jobs"

    id = Column(Integer, primary_key=True, index=True)
    dataset_id = Column(Integer, ForeignKey("training_datasets.id"), nullable=False, index=True)
    base_model_id = Column(String, nullable=False)
    status = Column(String, default="queued", nullable=False, index=True)
    task_id = Column(String, nullable=True, index=True)
    hyperparameters = Column(JSON, nullable=True)
    metrics = Column(JSON, nullable=True)
    output_path = Column(String, nullable=True)  # local path to adapter/model (LoRA)
    adapter_path = Column(String, nullable=True)  # remote storage path for adapter
    merged_path = Column(String, nullable=True)  # local merged HF model dir
    gguf_local_path = Column(String, nullable=True)
    artifact_path = Column(String, nullable=True)  # remote path for GGUF artifact
    gguf_path = Column(String, nullable=True)  # alias for remote gguf
    model_id = Column(String, nullable=True, index=True)
    validation_status = Column(String, default="pending", nullable=False)
    validation_metrics = Column(JSON, nullable=True)
    error = Column(Text, nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    started_at = Column(DateTime(timezone=True), nullable=True)
    completed_at = Column(DateTime(timezone=True), nullable=True)

    dataset = relationship("TrainingDataset", back_populates="jobs")


__all__ = ["TrainingDataset", "TrainingJob"]

File: frontend/src/api/training.ts
import apiClient from '../services/api'

export interface TrainingDataset {
  id: number
  name: string
  description?: string
  status: string
  total_examples: number
  created_at?: string
  updated_at?: string
  config?: Record<string, unknown>
  sample_preview?: Array<Record<string, unknown>>
}

export interface TrainingJob {
  id: number
  dataset_id: number
  base_model_id: string
  status: string
  metrics?: Record<string, unknown>
  adapter_path?: string
  artifact_path?: string
  gguf_path?: string
  model_id?: string
  validation_status?: string
  validation_metrics?: Record<string, unknown>
  created_at?: string
  started_at?: string
  completed_at?: string
  output_path?: string
}

export interface CreateDatasetPayload {
  name: string
  description?: string
  document_ids: number[]
  questions_per_chunk?: number
  total_pairs_target?: number
  qa_model_id?: string
  prompt_template?: string
}

export interface CreateJobPayload {
  dataset_id: number
  base_model_id: string
  model_id?: string
  hyperparameters?: Record<string, unknown>
  validation_prompts?: string[]
}

export interface TrainingJobArtifacts {
  job_id: number
  model_id?: string
  adapter_url?: string
  gguf_url?: string
}

export const trainingApi = {
  listDatasets: async (): Promise<TrainingDataset[]> => {
    const response = await apiClient.get<TrainingDataset[]>('/training/datasets')
    return response.data
  },
  createDataset: async (payload: CreateDatasetPayload): Promise<TrainingDataset> => {
    const response = await apiClient.post<TrainingDataset>('/training/datasets', payload)
    return response.data
  },
  listJobs: async (): Promise<TrainingJob[]> => {
    const response = await apiClient.get<TrainingJob[]>('/training/jobs')
    return response.data
  },
  createJob: async (payload: CreateJobPayload): Promise<TrainingJob> => {
    const response = await apiClient.post<TrainingJob>('/training/jobs', payload)
    return response.data
  },
  getJobArtifacts: async (jobId: number): Promise<TrainingJobArtifacts> => {
    const response = await apiClient.get<TrainingJobArtifacts>(`/training/jobs/${jobId}/artifacts`)
    return response.data
  },
}


